### PR TITLE
Add controlnets

### PR DESCRIPTION
**Pull request by jfischoff**
Adding the ability to add controlnet for sd and sdxl. I needed to modify the diffusers pipeline because I need to support optional controlnets and diffusers does not support that in a single pipeline.

### CODE REVIEW

File: text_to_image/model.py [hunk 1]
```diff
@@ -48,6 +54,46 @@ class Embedding(BaseModel):
...
+    end_percentage: float = Field(
+        default=1.0,
+        description="""
+            The percentage of the image to end applying the controlnet in terms of the total timesteps.
+        """,
+        ge=0.0,
+        le=1.0,
+    )
+
+
```
Comment (badayvedat) [inline text_to_image/model.py:96]: can you add a validator for checking start_percentage < end_percentage

File: text_to_image/model.py [hunk 2]
```diff
@@ -255,6 +317,29 @@ def generate_image(input: InputParameters) -> OutputParameters:
...
+                # download all the controlnet images
+                controlnet_images = []
+                for controlnet in input.controlnets:
+                    controlnet_image_path = download_file(
+                        controlnet.image_url, IMAGE_DIR
+                    )
+                    controlnet_image = PIL.Image.open(controlnet_image_path)
+                    controlnet_images.append(controlnet_image)
+
+                kwargs["image"] = controlnet_images
```
Comment (badayvedat) [inline text_to_image/model.py:341]: when do these control images are preprocessed (canny, depth, pose, etc.)

File: text_to_image/runtime.py [hunk 3]
```diff
@@ -16,11 +16,14 @@
...
 from pydantic import BaseModel, Field
 
+from text_to_image.pipeline import create_pipeline
+
 DeviceType = Literal["cpu", "cuda"]
 
 CHECKPOINTS_DIR = Path("/data/checkpoints")
 LORA_WEIGHTS_DIR = Path("/data/loras")
 EMBEDDINGS_DIR = Path("/data/embeddings")
+IMAGE_DIR = Path("/data/images")
```
Comment (badayvedat) [inline text_to_image/runtime.py:26]: why would we persist them?

File: text_to_image/runtime.py [hunk 4]
```diff
@@ -97,6 +100,46 @@ class Embedding(BaseModel):
     )
 
 
+class ControlNet(BaseModel):
+    path: str = Field(
+        description="URL or the path to the control net weights.",
+        examples=[
+            "https://civitai.com/api/download/models/135931",
+        ],
+    )
```
Comment (badayvedat) [inline text_to_image/runtime.py:109]: duplicate in `model.py`

File: text_to_image/runtime.py [hunk 5]
```diff
@@ -260,6 +302,70 @@ def get_model(self, model_name: str, arch: str) -> Model:
...
+            yield
+            return
+
+        controlnet_paths = []
+        for controlnet in controlnets:
+            try:
+                # see if you can parse the controlnet path as a URL
+                if controlnet.path.startswith("http://") or controlnet.path.startswith(
+                    "https://"
+                ):
```
Comment (badayvedat) [inline text_to_image/runtime.py:321]: let's limit to https

File: text_to_image/runtime.py [hunk 6]
```diff
@@ -260,6 +302,70 @@ def get_model(self, model_name: str, arch: str) -> Model:
...
+                422,
+                detail=f"Failed to load controlnet: {e}",
+            )
+
+        print(
+            "adding controlnets to the pipe, controlnet_models len",
+            len(controlnet_models),
+        )
+
+        pipe.controlnet = MultiControlNetModel(controlnet_models)
```
Comment (badayvedat) [inline text_to_image/runtime.py:365]: GPU operations should be handled by Model Manager to ensure that the application won't run into CUDA OOM

File: text_to_image/model.py (hunk 2 above)
Comment (jfischoff) [reply text_to_image/model.py:341]: Beforehand. The idea is that this is a base endpoint for use with workflows. The preprocessing happens in an earlier step in the workflow

File: text_to_image/runtime.py (hunk 3 above)
Comment (jfischoff) [reply text_to_image/runtime.py:26]: It should not be. Will change

File: text_to_image/runtime.py (hunk 4 above)
Comment (jfischoff) [reply text_to_image/runtime.py:109]: I'll remove the example

File: text_to_image/runtime.py (hunk 6 above)
Comment (jfischoff) [reply text_to_image/runtime.py:365]: What do you mean? I'm not seeing a Model Manager anywhere?

Comment (badayvedat) [reply text_to_image/runtime.py:365]: https://github.com/fal-ai/text-to-image/blob/main/text_to_image/runtime.py#L446-L483

see these example usages
https://github.com/fal-ai/text-to-image/blob/main/text_to_image/runtime.py#L379
https://github.com/fal-ai/text-to-image/blob/main/text_to_image/model.py#L262

File: text_to_image/model.py [hunk 7]
```diff
@@ -332,10 +360,11 @@ def generate_image(input: InputParameters) -> OutputParameters:
                 # download all the controlnet images
                 controlnet_images = []
                 for controlnet in input.controlnets:
-                    controlnet_image_path = download_file(
-                        controlnet.image_url, IMAGE_DIR
+                    # TODO replace with something that doesn't download the image every time
+                    controlnet_image = read_image_from_url(
+                        controlnet.image_url
```
Comment (badayvedat) [inline text_to_image/model.py]: you can use lru_cache for download function, so you it won't download the image every time

**Review by badayvedat — APPROVED** [review]

**Pull request merged**